File: clientFunctions/api_client.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <curl/curl.h>
#include <json/json.h>

class ApiClient {
private:
    CURL* curl = nullptr;
    struct curl_slist* headers = nullptr;
    std::string responseBuffer;

    static size_t writeCallback(void* contents, size_t size, size_t nmemb, void* userp) {
        ((std::string*)userp)->append((char*)contents, size * nmemb);
        return size * nmemb;
    }

    Json::Value parseJsonResponse(const std::string& response) {
        Json::CharReaderBuilder reader;
        Json::Value jsonResponse;
        std::string errors;
        std::istringstream stream(response);
        if (!Json::parseFromStream(reader, stream, &jsonResponse, &errors)) {
            std::cerr << "Failed to parse JSON response: " << errors << std::endl;
            return Json::Value();
        }
        return jsonResponse;
    }

public:
    ApiClient() {
        curl = curl_easy_init();
        if (!curl) {
            throw std::runtime_error("Failed to initialize curl.");
        }
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, ApiClient::writeCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &responseBuffer);
        headers = curl_slist_append(headers, "Content-Type: application/json");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    }

    ~ApiClient() {
        if (curl) curl_easy_cleanup(curl);
        if (headers) curl_slist_free_all(headers);
    }

    Json::Value execQuery(const std::string& query) {
        std::string url = "http://localhost:8080/api/execQuery";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, query.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, query.length());

        responseBuffer.clear();
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            std::cerr << "Curl error: " << curl_easy_strerror(res) << std::endl;
            return Json::Value();
        }
        return parseJsonResponse(responseBuffer);
    }

    Json::Value listTables() {
        std::string url = "http://localhost:8080/api/listTables";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPGET, 1L);

        responseBuffer.clear();
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            std::cerr << "Curl error: " << curl_easy_strerror(res) << std::endl;
            return Json::Value();
        }
        return parseJsonResponse(responseBuffer);
    }

    bool insert(const std::string& tableName, const std::string& values) {
        std::string jsonData = "{\"tableName\": \"" + tableName + "\", \"values\": " + values + "}";
        std::string url = "http://localhost:8080/api/insert";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonData.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, jsonData.length());

        responseBuffer.clear();
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            std::cerr << "Curl error: " << curl_easy_strerror(res) << std::endl;
            return false;
        }
        Json::Value response = parseJsonResponse(responseBuffer);
        return response["status"].asString() == "success";
    }

    bool deleteData(const std::string& tableName, const std::string& columns, const std::string& values) {
        std::string jsonData = "{\"tableName\": \"" + tableName + "\", \"columns\": " + columns + ", \"values\": " + values + "}";
        std::string url = "http://localhost:8080/api/delete";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonData.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, jsonData.length());

        responseBuffer.clear();
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            std::cerr << "Curl error: " << curl_easy_strerror(res) << std::endl;
            return false;
        }
        Json::Value response = parseJsonResponse(responseBuffer);
        return response["status"].asString() == "success";
    }

    Json::Value select(const std::string& tableName, const std::string& columns, const std::string& whereClause, const std::string& params) {
        std::string jsonData = "{\"tableName\": \"" + tableName + "\", \"columns\": " + columns + ", \"whereClause\": \"" + whereClause + "\", \"params\": " + params + "}";
        std::string url = "http://localhost:8080/api/select";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonData.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, jsonData.length());

        responseBuffer.clear();
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            std::cerr << "Curl error: " << curl_easy_strerror(res) << std::endl;
            return Json::Value();
        }
        return parseJsonResponse(responseBuffer);
    }
};

File: clientFunctions/ApiClient.java
import org.json.JSONObject;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class ApiClient {

    private final String baseUrl;
    private final HttpClient httpClient;

    public ApiClient(String baseUrl) {
        this.baseUrl = baseUrl;
        this.httpClient = HttpClient.newHttpClient();
    }

    private JSONObject sendRequest(HttpRequest request) {
        try {
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() == 200) {
                return new JSONObject(response.body());
            } else {
                System.err.println("HTTP request failed with code: " + response.statusCode());
                return null;
            }
        } catch (IOException | InterruptedException e) {
            System.err.println("Error during HTTP request: " + e.getMessage());
            return null;
        }
    }

    public JSONObject execQuery(String query) {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/execQuery"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString("{\"query\": \"" + query + "\"}"))
                .build();
        return sendRequest(request);
    }

    public boolean createTable(String tableSql) {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/createTable"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(tableSql))
                .build();
        JSONObject response = sendRequest(request);
        return response != null && "success".equals(response.optString("status"));
    }

    public JSONObject listTables() {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/listTables"))
                .GET()
                .build();
        return sendRequest(request);
    }

    public boolean insert(String tableName, String values) {
        String jsonData = String.format("{\"tableName\": \"%s\", \"values\": %s}", tableName, values);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/insert"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
        JSONObject response = sendRequest(request);
        return response != null && "success".equals(response.optString("status"));
    }

    public boolean delete(String tableName, String columns, String values) {
        String jsonData = String.format("{\"tableName\": \"%s\", \"columns\": %s, \"values\": %s}", tableName, columns, values);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/delete"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
        JSONObject response = sendRequest(request);
        return response != null && "success".equals(response.optString("status"));
    }

    public JSONObject select(String tableName, String columns, String whereClause, String params) {
        String jsonData = String.format(
            "{\"tableName\": \"%s\", \"columns\": %s, \"whereClause\": \"%s\", \"params\": %s}",
            tableName, columns, whereClause, params
        );
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/select"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
        return sendRequest(request);
    }
}

